pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+	pub(super) type ActiveSessionData<T: Config> = StorageValue<_, SessionData<T>, ValueQuery>;
+
+	/// Pending update data used when the current session finalizes.
+	/// Once it's used for the update, it's reset.
+	#[pallet::storage]
+	#[pallet::getter(fn next_session_changes)]
+	pub(super) type NextSessionChanges<T: Config> = StorageValue<_, SessionChanges<T>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		NewSession {
+			collator_reward: T::Balance,
+			total_reward: T::Balance,
+			last_changes: SessionChanges<T>,
+		},
+		SessionAdvancementFailed {
+			error: DispatchError,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Limit of max calls with same id to [`Pallet::set_collator_reward()`] or
+		/// [`Pallet::set_total_reward()`] reached.
+		MaxChangesPerSessionReached,
+		InsufficientTotalReward,
+	}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub collators: Vec<T::AccountId>,
+		pub collator_reward: T::Balance,
+		pub total_reward: T::Balance,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			GenesisConfig {
+				collators: Default::default(),
+				collator_reward: Default::default(),
+				total_reward: Default::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			T::Rewards::attach_currency(
+				(T::Domain::get(), T::StakeCurrency::get()),
+				COLLATOR_GROUP_ID,
+			)
+			.map_err(|e| log::error!("Failed to attach currency to collator group: {:?}", e))
+			.ok();
+
+			ActiveSessionData::<T>::mutate(|session_data| {
+				session_data.collator_count = self.collators.len().saturated_into();
+				session_data.collator_reward = self.collator_reward;
+				session_data.total_reward = self.total_reward;
+			});
+
+			// Enables rewards already in genesis session.
+			for collator in &self.collators {
+				Pallet::<T>::do_init_collator(collator)
+					.map_err(|e| {
+						log::error!("Failed to init genesis collators for rewards: {:?}", e);
+					})
+					.ok();
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Claims the reward the associated to a currency.
+		/// The reward will be transferred to the target account.
+		#[pallet::weight(T::WeightInfo::claim_reward())]
+		pub fn claim_reward(origin: OriginFor<T>, account_id: T::AccountId) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			T::Rewards::claim_reward((T::Domain::get(), T::StakeCurrency::get()), &account_id)
+				.map(|_| ())
+		}
+
+		/// Admin method to set the reward amount for a collator used for the next sessions.
+		/// Current session is not affected by this call.
+		#[pallet::weight(T::WeightInfo::set_collator_reward())]
+		pub fn set_collator_reward(
+			origin: OriginFor<T>,
+			collator_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
```

Comment:
I think this is not needed, but a good sanity check maybe as we also do this check in `set_total_reward`. Wdyt @wischli or is the check in `set_total_reward` otherwise a safety mesaure?

* `do_advance`: Already does check `let total_reward = min(num_collators*reward_collator, total_reward)`

Reply:
IMO we should have the check in both locations to minimize the chance of having less total rewards than needed for the collators in `do_advance`.

Reply:
[27e4653](https://github.com/centrifuge/centrifuge-chain/pull/1198/commits/27e46538b255be5e3ed457e97b60797270739fce)